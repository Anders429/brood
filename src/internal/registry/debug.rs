use crate::{
    component::Component,
    internal::{archetype::Archetype, entity::EntityDebug},
    registry::{NullRegistry, Registry},
};
use alloc::boxed::Box;
use core::{
    any::{type_name, Any},
    fmt::{Debug, DebugMap},
    marker::PhantomData,
};
use unsafe_any::UnsafeAnyExt;

pub trait RegistryDebug: Registry {
    unsafe fn debug<'a, 'b, E, R>(
        key: &[u8],
        index: usize,
        bit: usize,
        archetype: &Box<dyn Any>,
        debug_map: &mut DebugMap<'a, 'b>,
        entity: PhantomData<E>,
        registry: PhantomData<R>,
    ) where
        R: Registry,
        E: EntityDebug;
}

impl RegistryDebug for NullRegistry {
    unsafe fn debug<'a, 'b, E, R>(
        key: &[u8],
        index: usize,
        bit: usize,
        archetype: &Box<dyn Any>,
        debug_map: &mut DebugMap<'a, 'b>,
        entity: PhantomData<E>,
        registry: PhantomData<R>,
    ) where
        R: Registry,
        E: EntityDebug,
    {
        debug_map.entry(
            &type_name::<E>(),
            &archetype.downcast_ref_unchecked::<Archetype<E>>(),
        );
    }
}

impl<C, R1> RegistryDebug for (C, R1)
where
    C: Component + Debug,
    R1: RegistryDebug,
{
    unsafe fn debug<'a, 'b, E, R2>(
        key: &[u8],
        index: usize,
        bit: usize,
        archetype: &Box<dyn Any>,
        debug_map: &mut DebugMap<'a, 'b>,
        entity: PhantomData<E>,
        registry: PhantomData<R2>,
    ) where
        R2: Registry,
        E: EntityDebug,
    {
        let mut new_bit = bit + 1;
        let new_index = if bit >= 8 {
            new_bit %= 8;
            index + 1
        } else {
            index
        };

        if key.get_unchecked(index) & (1 << bit) != 0 {
            R1::debug::<(C, E), R2>(
                key,
                new_index,
                new_bit,
                archetype,
                debug_map,
                PhantomData,
                PhantomData,
            );
        } else {
            R1::debug::<E, R2>(
                key,
                new_index,
                new_bit,
                archetype,
                debug_map,
                PhantomData,
                PhantomData,
            );
        }
    }
}
