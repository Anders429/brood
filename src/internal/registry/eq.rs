use crate::{
    component::Component,
    internal::{archetype::Archetype, entity::EntityPartialEq},
    registry::{NullRegistry, Registry},
};
use alloc::boxed::Box;
use core::{any::Any, marker::PhantomData};
use unsafe_any::UnsafeAnyExt;

pub trait RegistryPartialEq: Registry {
    unsafe fn eq<E, R>(
        key: &[u8],
        index: usize,
        bit: usize,
        archetype_a: &Box<dyn Any>,
        archetype_b: &Box<dyn Any>,
        entity: PhantomData<E>,
        registry: PhantomData<R>,
    ) -> bool
    where
        E: EntityPartialEq,
        R: Registry;
}

impl RegistryPartialEq for NullRegistry {
    unsafe fn eq<E, R>(
        _key: &[u8],
        _index: usize,
        _bit: usize,
        archetype_a: &Box<dyn Any>,
        archetype_b: &Box<dyn Any>,
        _entity: PhantomData<E>,
        _registry: PhantomData<R>,
    ) -> bool
    where
        E: EntityPartialEq,
        R: Registry,
    {
        archetype_a.downcast_ref_unchecked::<Archetype<E>>()
            == archetype_b.downcast_ref_unchecked::<Archetype<E>>()
    }
}

impl<C, R1> RegistryPartialEq for (C, R1)
where
    C: Component + PartialEq,
    R1: RegistryPartialEq,
{
    unsafe fn eq<E, R2>(
        key: &[u8],
        index: usize,
        bit: usize,
        archetype_a: &Box<dyn Any>,
        archetype_b: &Box<dyn Any>,
        _entity: PhantomData<E>,
        _registry: PhantomData<R2>,
    ) -> bool
    where
        E: EntityPartialEq,
        R2: Registry,
    {
        let mut new_bit = bit + 1;
        let new_index = if bit >= 8 {
            new_bit %= 8;
            index + 1
        } else {
            index
        };

        if key.get_unchecked(index) & (1 << bit) != 0 {
            R1::eq::<(C, E), R2>(
                key,
                new_index,
                new_bit,
                archetype_a,
                archetype_b,
                PhantomData,
                PhantomData,
            )
        } else {
            R1::eq::<E, R2>(
                key,
                new_index,
                new_bit,
                archetype_a,
                archetype_b,
                PhantomData,
                PhantomData,
            )
        }
    }
}

pub trait RegistryEq: RegistryPartialEq {}

impl RegistryEq for NullRegistry {}

impl<C, R> RegistryEq for (C, R)
where
    C: Component + Eq,
    R: RegistryEq,
{
}
