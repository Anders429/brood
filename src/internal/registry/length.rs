use crate::{component::Component, registry::NullRegistry};

pub trait RegistryLength {
    const LEN: usize;
}

impl RegistryLength for NullRegistry {
    const LEN: usize = 0;
}

impl<C, R> RegistryLength for (C, R)
where
    C: Component,
    R: RegistryLength,
{
    const LEN: usize = R::LEN + 1;
}
