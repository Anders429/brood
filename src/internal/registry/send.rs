use crate::{
    component::Component,
    registry::{NullRegistry, Registry},
};

pub trait RegistrySend: Registry {}

impl RegistrySend for NullRegistry {}

impl<C, R> RegistrySend for (C, R)
where
    C: Component + Send,
    R: RegistrySend,
{
}
