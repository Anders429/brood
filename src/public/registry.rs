use crate::{component::Component, internal::registry::RegistrySeal};

#[derive(Clone, Copy, Debug, Default, Eq, Hash, Ord, PartialEq, PartialOrd)]
pub struct NullRegistry;

pub trait Registry: RegistrySeal {}

impl Registry for NullRegistry {}

impl<C, R> Registry for (C, R)
where
    C: Component,
    R: Registry,
{
}

#[macro_export]
macro_rules! registry {
    ($component:ty $(,$components:ty)* $(,)?) => {
        ($component, registry!($($components,)*))
    };
    () => {
        $crate::registry::NullRegistry
    };
}
