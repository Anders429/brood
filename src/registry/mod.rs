//! A heterogeneous list of registered [`Component`]s.
//!
//! [`Registry`]s are most often defined using the [`Registry!`] macro. The items contained within
//! this module should rarely be needed in user code.
//!
//! Recommended practice is to define a `Registry` as a custom type, and use that type when
//! defining a [`World`].
//!
//! # Example
//! ``` rust
//! use brood::{
//!     Registry,
//!     World,
//! };
//!
//! // Define components.
//! struct Foo(usize);
//! struct Bar(bool);
//!
//! type Registry = Registry!(Foo, Bar);
//!
//! let world = World::<Registry>::new();
//! ```
//!
//! [`Component`]: crate::component::Component
//! [`Registry`]: crate::registry::Registry
//! [`Registry!`]: crate::Registry!
//! [`World`]: crate::world::World

pub(crate) mod contains;

mod debug;
mod eq;
mod sealed;
#[cfg(feature = "serde")]
mod serde;

#[cfg(feature = "serde")]
pub use self::serde::{
    Deserialize,
    Serialize,
};
#[cfg(feature = "rayon")]
pub use contains::ContainsParQuery;
pub use contains::{
    ContainsComponent,
    ContainsEntities,
    ContainsEntity,
    ContainsQuery,
};
pub use debug::Debug;
pub use eq::{
    Eq,
    PartialEq,
};

#[cfg(feature = "rayon")]
pub(crate) use contains::ContainsParViews;
pub(crate) use contains::ContainsViews;
#[cfg(feature = "rayon")]
pub(crate) use sealed::CanonicalParViews;
pub(crate) use sealed::{
    Canonical,
    CanonicalViews,
    Length,
};

use crate::{
    component::Component,
    hlist::define_null_uninstantiable,
};
use sealed::Sealed;

define_null_uninstantiable!();

/// A heterogeneous list of [`Component`]s.
///
/// Registries are used when defining [`World`]s. In order for components to be stored within a
/// `World`, they must be included in the `World`'s registry. However, care should be made to only
/// include `Component`s that will be used, as unused `Component`s will cause unnecessary heap
/// allocations.
///
/// While duplicate `Component`s can be included within a registry, it is not advised. There are no
/// benefits to including multiple `Component`s, and the unused components cause higher memory
/// allocation within a `World`.
///
/// # Example
/// ``` rust
/// use brood::Registry;
///
/// // Define components.
/// struct Foo(usize);
/// struct Bar(bool);
///
/// type Registry = Registry!(Foo, Bar);
/// ```
///
/// [`Component`]: crate::component::Component
/// [`World`]: crate::World
pub trait Registry: Sealed + 'static {}

impl Registry for Null {}

impl<C, R> Registry for (C, R)
where
    C: Component,
    R: Registry,
{
}

/// Creates a registry from the provided components.
///
/// This macro allows a registry to be defined without needing to manually create a heterogeneous
/// list of components. Given an arbitrary number of components, this macro will arrange them into
/// nested tuples forming a heterogeneous list.
///
/// A registry is not normally instantiated. Its main purpose is to be used as a generic in the
/// definition of a [`World`].
///
/// # Example
/// ``` rust
/// use brood::{
///     Registry,
///     World,
/// };
///
/// // Define components `Foo` and `Bar`.
/// struct Foo(u16);
/// struct Bar(f32);
///
/// // Define a registry containing those components.
/// type Registry = Registry!(Foo, Bar);
///
/// // Define a world using the registry.
/// let world = World::<Registry>::new();
/// ```
///
/// [`World`]: crate::World
#[macro_export]
macro_rules! Registry {
    ($component:ty $(,$components:ty)* $(,)?) => {
        ($component, $crate::Registry!($($components,)*))
    };
    () => {
        $crate::registry::Null
    };
}
