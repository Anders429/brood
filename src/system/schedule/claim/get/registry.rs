pub struct Index;

pub trait Get<T, I> {
    type Remainder;
}

impl<T, U> Get<T, Index> for (T, U) {
    type Remainder = U;
}

impl<I, T, U, V> Get<T, (I,)> for (U, V)
where
    V: Get<T, I>,
{
    type Remainder = (U, <V as Get<T, I>>::Remainder);
}
