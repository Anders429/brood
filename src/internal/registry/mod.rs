mod length;
mod storage;

use crate::{component::Component, registry::NullRegistry};
use length::RegistryLength;
use storage::RegistryStorage;

pub trait RegistrySeal: RegistryLength + RegistryStorage {}

impl RegistrySeal for NullRegistry {}

impl<C, R> RegistrySeal for (C, R)
where
    C: Component,
    R: RegistrySeal,
{
}
