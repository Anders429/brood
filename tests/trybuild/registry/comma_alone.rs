use brood::registry;

fn main() {
    let registry = registry!(,);
}
