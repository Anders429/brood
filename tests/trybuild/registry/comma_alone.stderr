error: no rules expected the token `,`
 --> tests/trybuild/registry/comma_alone.rs:4:30
  |
4 |     let registry = registry!(,);
  |                              ^ no rules expected this token in macro call
